### PR TITLE
Memes

### PR DESCRIPTION
**Pull request by goodmind**

### CODE REVIEW

File: src/core/util/web.ts [hunk 1]
```diff
@@ -95,4 +94,30 @@ export class Web {
...
+      headers: {
+        'User-Agent': 'Mozilla/5.0',
+      },
+      method: 'GET',
+    });
+
+    return res.text();
+  }
+
+  get(url: string, options = { qs: {} }) {
```
Comment (alexbft) [inline src/core/util/web.ts]: specify return type

File: src/core/util/web_exception.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+export class WebException extends Error {
+  constructor(message?: string) {
+    super(message);
+    Object.setPrototypeOf(this, new.target.prototype);
+  }
+}
```
Comment (alexbft) [inline src/core/util/web_exception.ts]: newline at eof

File: src/plugins/src/bashim.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
+import { Input } from 'core/bot_api/input';
+import { Plugin } from 'core/bot_api/plugin';
+import { Inject } from 'core/di/injector';
+import { TgApi } from 'core/tg/tg_api';
+import { Web } from 'core/util/web';
+import { AllHtmlEntities as Entities } from 'html-entities';
+import * as iconv from 'iconv-lite';
+import { Message } from 'node-telegram-bot-api';
+
+const mapRandom = (page: string) => {
```
Comment (alexbft) [inline src/plugins/src/bashim.ts]: use function instead of lambda

File: src/plugins/src/bashim.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  const id = page.match(/bash.im\/quote\/(\d+)/)![1];
+  const text = page.match(/0;">([^]+?)<\' \+ \'\/div>/)![1];
+  return [
+    id,
+    text.replace(/<' \+ 'br>/g, '\n').replace(/<' \+ 'br \/>/g, '\n'),
+  ];
+};
+
+const mapSpecific = (id: string) => (page: string) => {
```
Comment (alexbft) [inline src/plugins/src/bashim.ts]: same, also no currying

File: src/plugins/src/bashim.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+export class BashimPlugin implements Plugin {
+  readonly name = 'bash.im';
+
+  constructor(
+    private api: TgApi,
+    private web: Web,
+    private entities: Entities,
+  ) {}
+
+  init(input: Input) {
```
Comment (alexbft) [inline src/plugins/src/bashim.ts]: return type

File: src/plugins/src/bashim.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    input.onText(/^\!(баш|bash)\b[\s]*(\d+)?/, this.onMessage, this.onError);
+  }
+
+  onError = (msg: Message) => {
```
Comment (alexbft) [inline src/plugins/src/bashim.ts]: return type

File: src/plugins/src/bashim.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    this.api.sendMessage({
+      chat_id: msg.chat.id,
+      text: 'Баш уже не тот...',
+    });
+  }
+
+  onMessage = async (msg: Message, match: RegExpExecArray) => {
```
Comment (alexbft) [inline src/plugins/src/bashim.ts]: return type

File: src/plugins/src/bashim.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    const id = match[2];
+    const res = id == null
+      ? this.web.getAsBrowser('http://bash.im/forweb/?u').then(mapRandom)
```
Comment (alexbft) [inline src/plugins/src/bashim.ts]: use await

File: src/plugins/src/iq_test.ts [hunk 9]
```diff
@@ -18,15 +18,15 @@ const iq = (s: string) => {
 
 @Injectable
 export class IqTestPlugin implements Plugin {
-  public readonly name = 'IQ Test';
+  readonly name = 'IQ Test';
 
   constructor(private api: TgApi) { }
 
-  public init(input: Input) {
+  init(input: Input) {
```
Comment (alexbft) [inline src/plugins/src/iq_test.ts]: return type

File: src/core/bot_api/input.ts [hunk 10]
```diff
@@ -16,10 +18,15 @@ export class InputImpl implements Input {
     }
   }
 
-  onText(regex: RegExp, handler: MessageHandler): void {
-    this.eventEmitter.on('text', (msg) => {
-      if (regex.test(msg.text)) {
-        handler(msg);
+  onText(regex: RegExp, handler: MessageHandler, errorHandler: ErrorHandler = (_, e) => { throw e; }): void {
```
Comment (alexbft) [inline src/core/bot_api/input.ts]: don't use errorHandler. This was a mistake. please use try { ... } catch in the plugin.

File: src/plugins/src/bashim.ts (hunk 8 above)
Comment (goodmind) [reply src/plugins/src/bashim.ts]: why

File: src/plugins/src/iq_test.ts (hunk 9 above)
Comment (goodmind) [reply src/plugins/src/iq_test.ts]: meme

File: src/core/bot_api/input.ts (hunk 10 above)
Comment (goodmind) [reply src/core/bot_api/input.ts]: meme

File: src/plugins/src/bashim.ts (hunk 6 above)
Comment (goodmind) [reply src/plugins/src/bashim.ts]: meme

File: src/plugins/src/bashim.ts (hunk 5 above)
Comment (goodmind) [reply src/plugins/src/bashim.ts]: meme

File: src/plugins/src/bashim.ts (hunk 4 above)
Comment (goodmind) [reply src/plugins/src/bashim.ts]: meme

File: src/plugins/src/bashim.ts (hunk 3 above)
Comment (goodmind) [reply src/plugins/src/bashim.ts]: meemeeeees

File: src/core/util/web_exception.ts (hunk 2 above)
Comment (goodmind) [reply src/core/util/web_exception.ts]: ok

File: src/core/bot_api/input.ts (hunk 10 above)
Comment (alexbft) [reply src/core/bot_api/input.ts]: can you revert this file?

File: src/core/config/keys.ts [hunk 11]
```diff
@@ -1 +1,3 @@
+export const GoogleKey = Symbol();
```
Comment (alexbft) [inline src/core/config/keys.ts]: Symbol('GoogleKey')

File: src/core/config/keys.ts [hunk 12]
```diff
@@ -1 +1,3 @@
...
+export const GoogleCX = Symbol();
```
Comment (alexbft) [inline src/core/config/keys.ts]: export const GoogleCx = Symbol('GoogleCx');
style note: аббревиатуры в составе сложных идентификаторов должны, как и остальные слова, начинаться с заглавной и состоять из строчных букв.

Use text in symbol constructors to distinguish them in logs.

File: src/core/tg/tg_api.ts [hunk 13]
```diff
@@ -20,6 +24,15 @@ export class TgApi {
     return this.tgClient.send('sendMessage', args);
   }
 
+  editMessageText({ message_id, chat }: Message, text: string, args: EditMessageTextOptions) {
```
Comment (alexbft) [inline src/core/tg/tg_api.ts]: add return type

File: src/core/util/web.ts [hunk 14]
```diff
@@ -1,14 +1,13 @@
-import * as http from 'http';
 import { Injectable } from 'core/di/injector';
 import { Environment } from 'core/environment/environment';
+import * as http from 'http';
+// import * as https from 'https';
```
Comment (alexbft) [inline src/core/util/web.ts]: remove commented line

File: src/plugins/src/bashim.ts [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+
+  onError = (msg: Message): void => {
+    this.api.sendMessage({
+      chat_id: msg.chat.id,
+      text: 'Баш уже не тот...',
+    });
+  }
+
+  onMessage = async (msg: Message, match: RegExpExecArray): Promise<void> => {
+    const id = match[2];
```
Comment (alexbft) [inline src/plugins/src/bashim.ts]: use try { ... } catch (e) { this.onError(msg); }

File: src/plugins/src/images.ts [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+  readonly name = 'Images';
+
+  constructor(
+    private web: Web,
+    private api: TgApi,
+    @InjectValue(GoogleKey) private googlekey: string,
+    @InjectValue(GoogleCX) private googlecx: string,
+  ) {}
+
+  init(input: Input) {
```
Comment (alexbft) [inline src/plugins/src/images.ts]: return type

File: src/plugins/src/images.ts [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+    input.onText(
+      /!(покажи|пик|пек|img|pic|moar|моар|more|еще|ещё)(?: (.+))?/,
+      this.onMessage,
+      this.onError,
+    );
+  }
+
+  sendInline(msg: Message, pic, picSet, txt) {
```
Comment (alexbft) [inline src/plugins/src/images.ts]: return type, arg types

File: src/plugins/src/images.ts [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+    this.api.sendMessage({
+      chat_id: msg.chat.id,
+      text: url,
+      reply_markup: {
+        inline_keyboard: context.keyboard,
+      },
+    });
+  }
+
+  updateInline({ msg, pic, keyboard }: Context) {
```
Comment (alexbft) [inline src/plugins/src/images.ts]: return type

File: src/plugins/src/images.ts [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+    this.api.editMessageText(msg, pic.link, {
+      reply_markup: {
+        inline_keyboard: keyboard,
+      },
+    });
+  }
+
+  search(txt: string, offset?: number) {
```
Comment (alexbft) [inline src/plugins/src/images.ts]: return type

File: src/plugins/src/images.ts [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+    return this.rawSearch(txt, 8, offset);
+  }
+
+  async onCallback(context: Context, cb, msg: Message) {
```
Comment (alexbft) [inline src/plugins/src/images.ts]: return type, arg types

File: src/plugins/src/images.ts [hunk 21]
```diff
@@ -0,0 +1,171 @@
...
+          picSet = picSet.concat(results);
+        }
+        pic = picSet[index];
+        this.updateInline(context);
+        // cb.answer('');
+        break;
+    }
+  }
+
+  onMessage = async (msg: Message, match: RegExpExecArray) => {
```
Comment (alexbft) [inline src/plugins/src/images.ts]: return type

File: src/plugins/src/images.ts [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+    let txt = match[2];
```
Comment (alexbft) [inline src/plugins/src/images.ts]: use try ... catch instead of passing onError

File: src/plugins/src/images.ts [hunk 23]
```diff
@@ -0,0 +1,171 @@
...
+  }
+
+  onError = ({ chat }: Message) => {
+    this.api.sendMessage({
+      chat_id: chat.id,
+      text: 'Поиск не удался...',
+    });
+  }
+
+  private async rawSearch(txt: string, rsz = 1, offset = 1) {
```
Comment (alexbft) [inline src/plugins/src/images.ts]: return type

File: src/plugins/src/translate.ts [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+@Inject
+export class TranslatePlugin implements Plugin {
+  readonly name = 'Translate';
+
+  constructor(
+    private api: TgApi,
+    private web: Web,
+  ) {}
+
+  async translate(src: string, dest: string, txt: string) {
```
Comment (alexbft) [inline src/plugins/src/translate.ts]: return type

File: src/plugins/src/translate.ts [hunk 25]
```diff
@@ -0,0 +1,94 @@
...
+      const evalFn = new Function(`return ${res}`);
+      const json: string[][] = evalFn();
+      return json[0].map((d) => d[0]).join('');
+    } catch (e) {
+      console.error(e);
+      return null;
+    }
+  }
+
+  init(input: Input) {
```
Comment (alexbft) [inline src/plugins/src/translate.ts]: return type

File: src/plugins/src/translate.ts [hunk 26]
```diff
@@ -0,0 +1,94 @@
...
+    input.onText(
+      /!(переведи|translate|перевод|расшифруй|tr)( [a-z]{2})?( [a-z]{2})?(?: ([^]+))?$/,
+      this.onMessage,
+      this.onError,
+    );
+  }
+
+  onMessage = async (msg: Message, match: RegExpMatchArray) => {
```
Comment (alexbft) [inline src/plugins/src/translate.ts]: return type

File: src/plugins/src/translate.ts [hunk 27]
```diff
@@ -0,0 +1,94 @@
...
+    let src: string;
+    let dest: string;
+    let text: string;
+
+    if (match[2] != null && match[3] == null) {
```
Comment (alexbft) [inline src/plugins/src/translate.ts]: use try ... catch instead of onError

**Pull request merged**